### PR TITLE
Frontend new

### PR DESCRIPTION
**Pull request by mkatsimpris**
The new frontend part (frontend_v2.py) outputs serially the signals for each component. In this stage in the testbench the block is fed 3 times, one for each component. 

TODO
1)Add 3 rams (block buffers) in order to store the block and read in each iteration.
2) Change the design of the color converted in order to output each color component for each iteration

@cfelton @Vikram9866

### CODE REVIEW

**Comment by cfelton** [comment]
@mkatsimpris for you todo, I don't think you need 3 RAMs, the frontend row buffer will need some extra control signals, it will need to hold a block so you can read it three times.

We need to do some architectural analysis and estimate the throughput and resources.

**Comment by cfelton** [comment]
@mkatsimpris this PR can't be merged, it has conflicts.

**Comment by mkatsimpris** [comment]
I will fix the conflicts Chris. This happens because I didnt pull the latest changes from the main repo

**Comment by mkatsimpris** [comment]
It can store a block and read it three times as longs as the inpus is stopped with a "stop sending signal". Then when it reads three times the block it can de-assert the "stop sending signal" right?

**Comment by mkatsimpris** [comment]
I am currently measuring the performance of the frontend with Vivado synthesis tool.

**Comment by mkatsimpris** [comment]
There are some problems in the design because there are inferred  some latches. Due to the fact that the outputs of the dct-2d and zig-zag are parallel, in order to become serial they must be hold for some time. That causes a lot of problems in the timing due to the instantiation of latches.

**Comment by cfelton** [comment]
> It can store a block and read it three times as longs as the inpus is stopped with a "stop sending signal". Then when it reads three times the block it can de-assert the "stop sending signal" right?

Yes, in general.  Lets put this flow-control in the input row-buffer.  The frontend will receive an image block when it is ready (again simply `ready`-`valid` flow control) and will indicate when to present the next image block.  The input row buffer will need to halt the flow or have enough buffering to "double buffer".

**Comment by Vikram9866** [comment]
@mkatsimpris . Are you storing the output data in a `ROM` or `FIFO` or streaming the data directly to `backend` ?

**Comment by mkatsimpris** [comment]
@vikram I am streaming them directly to backend. First the 64 signals for the Y component the for Cb then for Cr. Is there any problem for you?

**Comment by Vikram9866** [comment]
I made a Double FIFO(set of 2 fifo's) from which my back-end module takes data. For communicating with Front End Iam sending a ready signal. On seeing the signal you can fill the data in the FIFO. Is it ok for you ?

**Comment by mkatsimpris** [comment]
Why are you sending a ready signal? I output a data_valid signal so you can
process the data.

**Comment by Vikram9866** [comment]
Backend sends a `ready` signal if it can take an extra block in. If it does not send a `ready` signal `front-end` must halt the processing of `next block` and wait for the `ready` signal. Iam expecting this between blocks.

**Comment by cfelton** [comment]
@mkatsimpris @Vikram9866 , yes the downstream subblocks need to be able to stop the flow of data, hence the 'ready'-'valid' flow control.

**Comment by mkatsimpris** [comment]
Why the processing in the backend part requires the stop of the flow?
If the flow from the frontend part has to stop, then I think some FIFOs must inserted in the output of my part. Am I right?

**Comment by Vikram9866** [comment]
@mkatsimpris Using a `FIFO` Interface will help us to avoid long `combinational logics` between blocks.
(We can manage doing our task without FIFO, but there may be timing issues).
There may be chance that `comb_logic@output(Frontend) and comb_logic@in(Backend)` may not meet timing requirements. So, a FIFO or some timing element will be very good.

What i think is that the `ready-valid` flow must be some thing-like this :(I suggest this)

`frontend(source)->DoubleBuffer(sink)->DoubleBuffer(source)->Backend(sink).`

For our task a RAM will make, we wont be needing a FIFO.
@cfelton . Do I make sense ?

@mkatsimpris  If the backend fails or lags some where, then it do not make sense for front-end to send the data. So, we need to stop the flow.

**Comment by mkatsimpris** [comment]
In order to reduce the long combination logic, you have to use pipeline in
order to split the logic between two registers. This reduces a lot the
critical path, and the clock frequency could be higher. If you pipelined
correctly your design then there is no need for FIFOs. So I think the use
of FIFOs is irrelevant with the long combinational blocks. The correct use
of FIFOs is to assure the right flow in each component. @cfelton am I right?

Also, specifically in which module you have to stop the income flow in
order to process data?

**Comment by cfelton** [comment]
I read through the above quickly, I might be missing some of the
nuance in the comments.

@mkatsimpris, I don't know the reasons why the downstream (backend)
needs to stop the flow but it is reasonable that it might.
@Vikram9866 will need to comment on the specifics.  Bursty processing
is reasonable (if not unavoidable).

I agree with @mkatsimpris on the combinatorial comments, I don't
think the low-level circuit details are what is being discussed
(they might be a consequence ...).  The issue at hand is the time
required for each block to complete its processing task and the
possible data-flow mismatch between blocks.

I updated the [JPEGEnc data flow](http://discourse.myhdl.org/t/jpeg-encoder-data-flow/94/12?u=cfelton) thread on discourse.  Attempting to capture the current 
state of the flow through the encoder.

**Comment by mkatsimpris** [comment]
Dont merge this PR. There are still some changes to be made.
1)Make some changes in the tests and pytest options
2)Change a little the color convertion v2

**Comment by cfelton** [comment]
@mkatsimpris ok, let me know when it is ready and I will review it.

File: .travis.yml [hunk 1]
```diff
@@ -43,7 +43,8 @@ install:
 
 script:
    - cd test
-   - coverage run --source=jpegenc -m py.test --include-reference
```
Comment (cfelton) [inline .travis.yml]: see the comments in PR #33.

Comment (mkatsimpris) [inline .travis.yml]: I have seen them Chris. I am trying to fix them. Today I will fix it.

Comment (cfelton) [inline .travis.yml]: @mkatsimpris no problem, I realize this is a work in progress, I wanted to make sure that these comments were not lost.

File: jpegenc/testing/testing.py [hunk 2]
```diff
@@ -1,25 +1,28 @@
...
         not pytest.config.getoption("--include-reference"),
         reason="reference tests, needs --include-reference option to run"
     )
 
+skipif_no_sim = pytest.mark.skipif(False, reason="simulator is not available")
+if hasattr(sys, '_called_from_test'):
+    skip_if_no_sim = pytest.mark.skipif(
+        True,
+        reason="simulator is not ok "
+    )
```
Comment (mkatsimpris) [inline jpegenc/testing/testing.py]: @cfelton  i added a new marker skip_if_no_sim. I set the _called_from_test attribute to true in the conftest but the test executes and it never skips. I want to put there as a condition the global variable sim_is_ok which is set in the conftest.py if the simulators are available. Do you know how to fix this?

File: jpegenc/subblocks/common/interfaces.py [hunk 3]
```diff
@@ -38,6 +46,28 @@ def __init__(self, nbits=8):
...
+
+    def __init__(self, nbits=8):
+        """member variables initialize"""
+        self.nbits = nbits
+        self.red = Signal(intbv(0)[nbits:])
+        self.green = Signal(intbv(0)[nbits:])
+        self.blue = Signal(intbv(0)[nbits:])
+        self.color_mode = Signal(intbv(0, min=0, max=3))
+        self.data_valid = Signal(bool(0))
+
```
Comment (mkatsimpris) [inline jpegenc/subblocks/common/interfaces.py]: There is no need for a seperate interface. I will clean up the code

**Comment by cfelton** [comment]
@mkatsimpris this pull-request is out of sync with the base, correct?  A lot of the changes that were made are either undone in this PR or were not merged correctly.

Example, in PR #22 a lot of things were re-organized, example moving the dct_1d and dct_2d to the same directory dct.  In this PR that is undone or was not merged?   Can you make a PR that only has your frontend changes and not other changes (e.g. the simulation options, reorg, etc.).  It will be much easier to review your buffer addition and other changes.  Thanks!, chris

**Comment by mkatsimpris** [comment]
@cfelton it isn't out of sync this request. By mistake there are the two folders dct_1d and dct_2d. I will remove them.

**Comment by mkatsimpris** [comment]
@cfelton  have you checked this PR?

File: jpegenc/subblocks/frontend/frontend_v2.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+from jpegenc.subblocks.zig_zag import zig_zag_scan, zig_zag
+from jpegenc.subblocks.common import YCbCr_v2, input_interface, outputs_2d, RGB_v2, outputs_frontend_new
+
+
+def frontend_transform(blockr, blockg, blockb, N=8):
+    """software implementation of the frontend part"""
+    ycbcr_blocks = [[[] for _ in range(N)] for _ in range(3)]
+    dct_blocks, dct_blocks_linear, zig_zag_blocks = [[] for _ in range(3)]
+
+    """Color space conversion"""
```
Comment (cfelton) [inline jpegenc/subblocks/frontend/frontend_v2.py]: Using the multiline strings for comments `""" comment """` will have a negative impact on run-time (albeit small).  In python the `#` is used for comments and yes python doesn't have a multi-line comment escape.  The exception is the doc-strings, where the multiline string right after a function definition can be parsed by various tools (e.g. the python interrupter / interactive shell).

No need to change all the string comments now but something to be aware of (I should have commented on this in earlier PRs).

**Pull request merged**